src/aotcompile.cpp

```diff
@@ -507,6 +503,922 @@ static void injectCRTAlias(Module &M, StringRef name, StringRef alias, FunctionT
+    }
+
+    timers.optimize.stopTimer();
+
+    if (opt) {
+        timers.opt.startTimer();
+        raw_string_ostream OS(*outputs);
+        PassBuilder PB;
+        AnalysisManagers AM{*TM, PB, OptimizationLevel::O0};
+        ModulePassManager MPM;
+        MPM.addPass(BitcodeWriterPass(OS));
+        *opt = NewArchiveMember(MemoryBufferRef(*outputs, names[1]));
+        outputs++;
+        timers.opt.stopTimer();
+    }
+
+    if (obj) {
+        timers.obj.startTimer();
+        SmallVector<char, 0> Buffer;
+        raw_svector_ostream OS(Buffer);
+        legacy::PassManager emitter;
+        addTargetPasses(&emitter, TM->getTargetTriple(), TM->getTargetIRAnalysis());
+        if (TM->addPassesToEmitFile(emitter, OS, nullptr, CGFT_ObjectFile, false))
+            jl_safe_printf("ERROR: target does not support generation of object files\n");
+        emitter.run(M);
+        *outputs = { Buffer.data(), Buffer.size() };
+        *obj = NewArchiveMember(MemoryBufferRef(*outputs, names[2]));
+        outputs++;
+        timers.obj.stopTimer();
+    }
+
+    if (asm_) {
+        timers.asm_.startTimer();
+        SmallVector<char, 0> Buffer;
+        raw_svector_ostream OS(Buffer);
+        legacy::PassManager emitter;
+        addTargetPasses(&emitter, TM->getTargetTriple(), TM->getTargetIRAnalysis());
+        if (TM->addPassesToEmitFile(emitter, OS, nullptr, CGFT_AssemblyFile, false))
+            jl_safe_printf("ERROR: target does not support generation of assembly files\n");
+        emitter.run(M);
+        *outputs = { Buffer.data(), Buffer.size() };
+        *asm_ = NewArchiveMember(MemoryBufferRef(*outputs, names[3]));
+        outputs++;
+        timers.asm_.stopTimer();
+    }
+}
+
+// serialize module to bitcode
+static auto serializeModule(const Module &M) {
+    assert(!verifyModule(M, &errs()) && "Serializing invalid module!");
+    SmallVector<char, 0> ClonedModuleBuffer;
+    BitcodeWriter BCWriter(ClonedModuleBuffer);
+    BCWriter.writeModule(M);
+    BCWriter.writeSymtab();
+    BCWriter.writeStrtab();
+    return ClonedModuleBuffer;
+}
+
+// Modules are deserialized lazily by LLVM, to avoid deserializing
+// unnecessary functions. We take advantage of this by serializing
+// the entire module once, then deleting the bodies of functions
+// that are not in this partition. Once unnecesary functions are
+// deleted, we then materialize the entire module to make use-lists
+// consistent.
+static void materializePreserved(Module &M, Partition &partition) {
+    DenseSet<GlobalValue *> Preserve;
+    for (auto &GV : M.global_values()) {
+        if (!GV.isDeclaration()) {
+            if (partition.globals.count(GV.getName())) {
+                Preserve.insert(&GV);
+            }
+        }
+    }
+    for (auto &F : M.functions()) {
+        if (!F.isDeclaration()) {
+            if (!Preserve.contains(&F)) {
+                F.deleteBody();
```

Comment:
I am not sure this results in legal IR, since there may still be attributes attached to the module for the declaration (such as [DISubprogram](https://llvm.org/doxygen/classllvm_1_1DISubprogram.html) and llvm.dbg.cu) which should not be added there. We used to have a function (prepare_call) for getting a proper declaration for a definition, but I think we removed that at some point. I thought maybe OrcJIT would have it, but it doesn't look like it includes support for handling debug information correctly either: https://llvm.org/doxygen/CompileOnDemandLayer_8cpp_source.html

Reply:
Can we just drop all metadata on the function? Since we're still attaching metadata to the real definition the linker might save us right?

Reply:
We probably don't need the metadata, though note this is module metadata, not on the function

Reply:
Module metadata seems more complicated since we're not allowed to drop module flags. However, we could go back to using CloneModule either before or after serialization, at a time cost. Presumably CloneModule should handle the debug info correctly.

Reply:
We emitted them into separate modules to avoid this issue. Can we keep them separate?

Reply:
It requires a bit more care to keep the modules separate since GPUCompiler also uses this codepath and probably assumes a single module, which I'm more hesitant to mess with.

For DISubprogram, the [LangRef](https://llvm.org/docs/LangRef.html#disubprogram) says it's legal to have it on a declaration for callsite debug info, so I think it's valid to leave it there despite deleting the function body. I suspect the same reasoning to apply to global variable declarations, though the LangRef does not mention that, but stripping them also shouldn't affect correctness. llvm.dbg.cu doesn't have much documentation, but I don't see anything obvious that LLVM doesn't already do (https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Utils/CloneFunction.cpp#L280 copies all the compile units, irrespective of if they're related to the function that was just cloned) so I'm not really seeing anything obviously wrong here.

Reply:
> It requires a bit more care to keep the modules separate since GPUCompiler also uses this codepath and probably assumes a single module, which I'm more hesitant to mess with.

We could merge it in GPUCompiler.jl, generally I would be in favor for exposing the work-list to GPUC, since we might want to implement caching differently for Enzyme vs CUDA